### PR TITLE
Night Statistics Monitoring

### LINKED ISSUE #700: Night Statistics Monitoring
**Description**

We want to develop and improve our product and, of course, there are some changes and its combinations which, according to some statistics, can make UTBot worse.

The main idea is collecting statistics after made changes. But it takes too long to collect statistics on a huge project to do it after each push into master. Thus, we will do it every night when no one makes changes. 

Also we need to visualize collected statistics to easily analyze them.

### PR DESCRIPTION
**Pull request by SBOne-Kenobi**
# Description

## Problem
We want to develop and improve our product and, of course, there are some changes and its combinations which, according to some statistics, can make UTBot worse.

## Solution
The main idea is collecting statistics after made changes. But it takes too long to collect statistics on a huge project to do it after each push into master. Thus, we will do it every night when no one makes changes. Also we need to visualize collected statistics to easily analyze them. 

Look at docs/NightStatisticsMonitoring.md for more information.

### Coverage
I decided to collect coverage information using our ConcreteExecutor and I've provided total instructions count from tested class.

Fixes #700 

## Type of Change

Please delete options that are not relevant.
- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

## Manual Scenario 

Created cloned repository and run github action. Also run each script with different parameters.

# Checklist:

- [x] The change followed the style guidelines of the UTBot project
- [x] Self-review of the code is passed
- [x] The change contains enough commentaries, particularly in hard-to-understand areas
- [x] New documentation is provided or existed one is altered

### CODE REVIEW

File: utbot-junit-contest/src/main/kotlin/org/utbot/contest/Statistics.kt [hunk 1]
```diff
@@ -51,7 +114,12 @@ class StatsForClass {
     val methodsWithAtLeastOneException: Int get() = statsForMethods.count { it.failReasons.isNotEmpty() }
     val testcasesGenerated: Int get() = statsForMethods.sumBy { it.testsGeneratedCount }
 
-    var coverage: CoverageInfo? = null
+    var coverage: Coverage? = null
+    var fuzzedCoverage: Coverage? = null
+    var concolicCoverage: Coverage? = null
```
Comment (sergeypospelov) [inline utbot-junit-contest/src/main/kotlin/org/utbot/contest/Statistics.kt:119]: Do we really need to have `Coverage` here as lists? Maybe it's OK to have only sets here?

**Review by sergeypospelov — CHANGES_REQUESTED** [review]
PR looks really nice, can't wait to try it out. There are just some minor things. Also here some thoughts about possible enhancements:
- What about adding `usedConcreteExecutors` field in the future? 😊
- Maybe in the future we could write an interface to extend statistics with a custom statistic produced from `UtExecution`?
- What about adding a special file from which the `StatisticsMonitoringKt::main` takes arguments? It could be useful for defining custom monitorings without modifying .yml file.

File: utbot-junit-contest/src/main/kotlin/org/utbot/monitoring/StatisticsMonitoring.kt [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+private val javaHome = System.getenv("JAVA_HOME")
+private val logger = KotlinLogging.logger {}
+
+fun main(args: Array<String>) {
+    val methodFilter: String?
+    val projectFilter: List<String>?
+    val processedClassesThreshold = 9999
+    val tools: List<Tool> = listOf(Tool.UtBot)
+    val timeLimit = 20
```
Comment (sergeypospelov) [inline utbot-junit-contest/src/main/kotlin/org/utbot/monitoring/StatisticsMonitoring.kt:27]: Why did we hardcode it?

File: utbot-junit-contest/src/main/kotlin/org/utbot/monitoring/StatisticsMonitoring.kt [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+    require(args.size == 3) {
+        "Wrong arguments: <output json> <run tries> <run timeout min> expected, but got: ${args.toText()}"
+    }
+
+    projectFilter = listOf("guava")
```
[the remark this answers is not in the document]
Comment (SBOne-Kenobi) [reply utbot-junit-contest/src/main/kotlin/org/utbot/monitoring/StatisticsMonitoring.kt:33]: No, I decided to collect statistics from only one project witch is used by me to test features. You, of course, can change it.

File: docs/NightStatisticsMonitoring.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+### Transforming, aggregating and rendering statistics
+Transforming adds total coverage statistics and timestamp.
+After that all collected statistics are aggregated by average function.
+Then history updates by aggregated statistics and rendered into 2 pictures:
+- coverage graph - graph with coverage statistics.
+- quantitative graph - graph with other quantitative statistics.
+
+
+Script: draw_stats_graphs.py.
+Input arguments: `<history file> <new stats file> <output coverage graph> <output quantitative graph>`.
```
[the remark this answers is not in the document]
Comment (SBOne-Kenobi) [reply docs/NightStatisticsMonitoring.md:87]: My mistake, I use our repository as database for statistics history. Format: json array of objects with statistics.

File: utbot-junit-contest/src/main/kotlin/org/utbot/contest/Statistics.kt (hunk 1 above)
Comment (SBOne-Kenobi) [reply utbot-junit-contest/src/main/kotlin/org/utbot/contest/Statistics.kt:119]: You are right, I decided to reuse `Coverage` class to don't create one more "Coverage".

File: monitoring/draw_stats_graphs.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    new_stats["timestamp"] = round(time() * MILLIS_IN_SEC)
+
+    return new_stats
+
+
+def update_stats_history(history_file, new_stats_file):
+    history = load(history_file) or []
+    new_stats = load(new_stats_file)
+    if new_stats is None:
+        raise FileNotFoundError("File with new stats not exists!")
```
Comment (victoriafomina) [inline monitoring/draw_stats_graphs.py:59]: does not exist :)

File: monitoring/draw_stats_graphs.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+from datetime import datetime
+from os.path import exists
+from sys import argv
+from time import time
+from collections import defaultdict
+
+import matplotlib.pyplot as plt
+from matplotlib.dates import DayLocator, ConciseDateFormatter
+
+DPI = 108
```
Comment (denis-fokin) [inline monitoring/draw_stats_graphs.py:11]: Could you add some clarification on these parameters? If them are parameters for some known library, please add a reference to the documentation.

File: utbot-junit-contest/src/main/kotlin/org/utbot/contest/Contest.kt [hunk 7]
```diff
@@ -216,19 +223,6 @@ fun runGeneration(
             generateWarningsForStaticMocking = false
         )
 
-    // Doesn't work
```
Comment (denis-fokin) [inline utbot-junit-contest/src/main/kotlin/org/utbot/contest/Contest.kt:219]: If the code does not work, why we include it into the change?

Comment (SBOne-Kenobi) [reply utbot-junit-contest/src/main/kotlin/org/utbot/contest/Contest.kt:219]: I just deleted an irrelevant and useless comment.

File: utbot-junit-contest/build.gradle [hunk 8]
```diff
@@ -92,4 +92,25 @@ jar { dependsOn classes
 
     duplicatesStrategy = DuplicatesStrategy.EXCLUDE
 
-}
\ No newline at end of file
+}
+
+task monitoringJar(type: Jar) { dependsOn classes
+
+    archiveName 'monitoring.jar'
```
Comment (denis-fokin) [inline utbot-junit-contest/build.gradle:99]: What about moving the value in gradle.properties?

Comment (SBOne-Kenobi) [reply utbot-junit-contest/build.gradle:99]: I think that name of jar file doesn't match to global information in gradle.properties.

**Comment by denis-fokin** [comment]
There are some suggestions on the approach to gather and store the statistics.

### Where to store
It is better to store the statistical data in a separate branch or even repository. This way we can update the data without any harm to the `main` branch.

### Metadata
Currently, there is not enough metadata for further data processing. Assuming, we store data in a separate repository or branch we may need the next data:

- commit hash
- build number
- environment

### Images
Maybe our way of the results representation is not the best, so we should not stick to the particular images. Images are still useful for brief insight into the specific data, but they should not be stored as bitmap images. It is better to choose some vector format.

The way of representation of overall data with a single picture is a separate challenge. By the way, an independent repository will help us with refining the images.

### Data processing
Eventually, it is great to have only data in the repository. There are several solutions on the data postprocessing:

- script in the repository
- service that is able to process the data and build visual representation of it

### Granularity
We need to choose a smallest piece of information that we want to represent in our statistics. It seems reasonable to use a test as such an atomic measure. A common scenario that a user can run into is a change of generated tests number. This will lead to the change of the results. In such situations it is good to see the reason right from the statistical data.

### Third party solutions
We can take a look at other's solutions:

File: utbot-junit-contest/src/main/kotlin/org/utbot/contest/Statistics.kt [hunk 9]
```diff
@@ -106,5 +176,20 @@ class FailReason(private val throwable: Throwable) {
...
+data class CoverageStatistic(val covered: Int, val total: Int)
+
+private fun CoverageInstructionsSet?.getCoverageInfo(className: String): CoverageStatistic = this?.run {
+    CoverageStatistic(
+        coveredInstructions.filter {
+            instr -> instr.className.startsWith(className)
+        }.toSet().size,
+        totalInstructions.toInt()
+    )
+} ?: CoverageStatistic(covered = 0, total = 0)
```
Comment (sergeypospelov) [inline utbot-junit-contest/src/main/kotlin/org/utbot/contest/Statistics.kt:195]: Still can't determine if it includes anonymous, nested, inner classes or not.

Comment (SBOne-Kenobi) [reply utbot-junit-contest/src/main/kotlin/org/utbot/contest/Statistics.kt:195]: Coverage from concrete executor doesn't calculate total instructions count from nested or inner classes. But we want to collect coverage in CoverageStatistic including them (that's a specific of ContestEstimator).

Comment (sergeypospelov) [reply utbot-junit-contest/src/main/kotlin/org/utbot/contest/Statistics.kt:195]: I mean, why didn't you add comments on that?

File: .github/workflows/night-statistics-monitoring.yml [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+          path: ${{ env.data_path }}
+
+      - name: Checkout aggregated monitoring data
+        uses: actions/checkout@v3
+        with:
+          ref: ${{ env.aggregated_data_branch }}
+          path: ${{ env.aggregated_data_path }}
+
+      - uses: actions/setup-java@v3
+        with:
```
Comment (victoriafomina) [inline .github/workflows/night-statistics-monitoring.yml:36]: It's better to run everything in the container as it'd give us the environment reproducible locally:
```
    runs-on: ubuntu-20.04
    container: unittestbot/java-env:java11-zulu-jdk-fx-gradle7.4.2-kotlinc1.7.0
    steps:
```

In this case KOTLIN_HOME env must be removed as it's already set up in the container

File: monitoring/build_aggregated_data.py [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+import json
+from os import listdir
+from os.path import isfile, join
+from time import time
+
+from monitoring_settings import JSON_VERSION
+from utils import *
+
+
+def get_file_seq(input_data_dir):
```
Comment (denis-fokin) [inline monitoring/build_aggregated_data.py:11]: Can we add type annotations?

**Review by denis-fokin — APPROVED** [review]

**Pull request merged**